transport_plugins/bled112/iotile_transport_bled112/broadcast_v2_dedupe.py

```diff
@@ -0,0 +1,104 @@
+"""This module is used to identify and filter out broadcast v2 broadcasts, which leads to significant
+performance increases.
+"""
+
+import time
+import struct
+from typing import Dict
+
+from iotile.cloud.utilities import device_id_to_slug
+
+def packet_is_broadcast_v2(packet: bytearray) -> bool:
+    """Simple/efficient check for whether a given packet from the bled112 is an IOTile Broadcast v2 packet."""
+    #Broadcast packets consist of 32 bytes for data, 10 for BLE packet header and 4 for bled112 bgapi header
+    if len(packet) != 46:
+        return False
+    #This identifies the bgapi packet as an event
+    if not (packet[0] == 0x80 and packet[2] == 6 and packet[3] == 0):
+        return False
+    #This identifies the event as a broadcast v2 packet
+    if not (packet[18] == 0x1b and packet[19] == 0x16 and packet[20] == 0xdd and packet[21] == 0xfd):
+        return False
+    return True
+
+class BroadcastV2DeduperCollection():
+    """Main interface into the Broadcast v2 deduplication code.
+
+    This contains a dictionary, keyed on the broadcast sender's encoded UUID, and with the values being
+    a small class that stores the last received packet from that UUID and the last time the packet
+    was forwarded. That class (bc_v2_deduper) will report whether the packet is new and should be allowed through.
+
+    Args:
+        pass_packets_every(float, seconds): For each encoded_uuid address, at least one packet will be allowed through
+            every "pass_packets_every" seconds
+    """
+
+    MAX_DEDUPERS = 500
+
+    def __init__(self, pass_packets_every: float = 5):
+        self._pass_packets_every = pass_packets_every
+        self.dedupers = {}  #type: Dict[bytes, BroadcastV2Deduper]
+
+    def allow_packet(self, packet: bytearray) -> bool:
+        """Run a packet through the broadcast_v2 deduper.
+
+        Returns False if the packet should be dropped
+        """
+
+        if not packet_is_broadcast_v2(packet):
+            return True
+
+        encoded_uuid = bytes(packet[22:26])
+        data = bytes(packet[22:])
+
+        deduper = self.dedupers.get(encoded_uuid)
+        if deduper is None:
+            deduper = BroadcastV2Deduper(encoded_uuid, self._pass_packets_every)
+            if len(self.dedupers) == self.MAX_DEDUPERS:
+                self.evict_oldest_deduper()
+            self.dedupers[encoded_uuid] = deduper
+
+        return deduper.allow_packet(data)
+
+    def evict_oldest_deduper(self):
+        """Find and remove the oldest deduper
+
+        This function will likely be called rarely, if ever
+        """
+        oldest = list(self.dedupers.values())[0]
+
+        for deduper in iter(self.dedupers.values()):
```

Comment:
`iter` is not needed here:

```
for deduper in self.dedupers.values():
```

Reply:
Actually given that the goal of this function is just to ensure the dict doesn't grow without bound, I would recommend a simpler eviction policy:

```python
def __init__(self):
    self.dedupers = OrderedDict()

def evict_oldest_deduper(self):
    first = next(iter(self.dedupers))  # From https://stackoverflow.com/a/21067850/9739119
  
    # Always delete the old pod that was added rather than finding out which pod is the least recently seen
    # if the pod is still active, it will come back at the end of the list and eventually the top of the dict will
    # be those pods that are seen rarely.
    del self.dedupers[first]
```

Reply:
Ok, sure.

Reply:
Got a test failure from that, looks like insertion order wasn't guaranteed until Python 3.6. Using the `OrderedDict` that Dima suggested instead.

Reply:
Mostly unimportant side info: technically we aren't *supposed* to rely on inherent dictionary ordering until Python 3.7 . It's considered an "implementation detail" in 3.6, so other versions of Python aren't guaranteed to support it (i.e. `pypy` or `Jython` might not compile that same requirement in to their interpreter/language as `Cython`, the reference implementation).

Interestingly, some places claim insertion order to be maintained all the way back to 3.5, but I more commonly see it cited in 3.6 .

http://gandenberger.org/2018/03/10/ordered-dicts-vs-ordereddict/ exemplifies the behavior difference in `IronPython` and even `pandas`. 

There is more gory discussion about this on the python-dev mailing list, https://mail.python.org/pipermail/python-dev/2016-September/146327.html . It was officially made a core feature rather than an implementation detail later, https://mail.python.org/pipermail/python-dev/2017-December/151283.html

Anyway, this comment is mostly just fun reading if anyone was curious about what is happening under the hood here :)